config/internal/configsource/manager.go

```diff
@@ -305,6 +313,109 @@ func (m *Manager) expandConfigSources(ctx context.Context, s string) (interface{
 	return retrieved.Value(), nil
 }
 
+// expandString expands environment variables and config sources that are specified on the string.
+func (m *Manager) expandString(ctx context.Context, s string) (interface{}, error) {
+	// Code based on os.Expand function. All delimiters that are checked against are
+	// ASCII so bytes are fine for this operation.
+	var buf []byte
+
+	// Using i, j, and w variables to keep correspondence with os.Expand code.
+	// i tracks the index in s from which a slice to be appended to buf should start.
+	// j tracks the char being currently checked and also the end of the slice to be appended to buf.
+	// w tracks the number of characters being consumed after a prefix identifying env vars or config sources.
+	i := 0
+	for j := 0; j < len(s); j++ {
+		if s[j] == expandPrefixChar && j+1 < len(s) {
+			if buf == nil {
+				buf = make([]byte, 0, 2*len(s))
+			}
+			buf = append(buf, s[i:j]...)
+
+			var retrieved interface{}
+			var err error
+			w := 0 // number of bytes consumed on this pass
+
+			switch {
+			case s[j+1] == expandPrefixChar:
+				// Escaped prefix:
+				buf = append(buf, s[j])
+				w = 1 // consumed a single char
+
+			case s[j+1] == '{':
+				// Bracketed usage, consume everything until first '}' exactly as os.Expand.
+				var content string
+				content, w = getShellName(s[j+1:])
+				content = strings.Trim(content, " ") // Allow for some spaces.
+				if len(content) > 1 && content[0] == expandPrefixChar {
+					name, _ := getShellName(content[1:])
+					if cfgSrc, ok := m.configSources[name]; ok {
+						// The name matches a config source, try to apply it.
+						cfgSrcInovke := expandEnvVars(content[1:])
+						retrieved, err = m.expandConfigSource(ctx, cfgSrc, cfgSrcInovke)
+						if err != nil {
+							return nil, err
+						}
+					}
+				}
+
+				if retrieved == nil {
+					// Not a config source, expand as os.ExpandEnv
+					buf = osExpandEnv(buf, content, w)
+				} else {
+					consumedAll := j+w+1 == len(s)
+					if consumedAll && len(buf) == 0 {
+						// This is the only content on the string, config
+						// source is free to return interface{}.
+						return retrieved, nil
+					}
+
+					// Either there was a prefix already or there are still
+					// characters to be processed.
+					buf = append(buf, fmt.Sprintf("%v", retrieved)...)
+				}
+
+			default:
+				// Non-bracketed usage, ie.: found the prefix char, it can be either a config
+				// source or an environment variable.
+				var name string
+				name, w = getShellName(s[j+1:])
+				if cfgSrc, ok := m.configSources[name]; ok {
```

Comment:
Same here. I am not sure it is a good idea to try a config source name and fall back to env variable name.
I think a better approach would be to differentiate config sources from env variables based on syntax. Can we do this:
- $name is always an env var if it is not followed by a `:`.
- $name:select is always a config source. `:` likely should have never be present in env variable names (if it was then it will be a breaking change, but it will be a one-time breaking change, which does not change behaviour when new config sources are introduced.
- ${name} is always an env var, assuming `:` character is not present in env var names. We don't support config sources with this syntax.
- ${name:...} is always a config source.

This maintains always full backwards compatibility for configs that used env variables. We may break configs which used env variables which contained `:` character but I think it is fine since it is likely very rare occurrence. We also do not allow using env variables that have `:` in the name at all going forward.

Will this work?

Reply:
The only case that I think is problematic is `$name:select` if you have something like `$HOMEDRIVE:\Windows\Temp` if we just allow the bracketed form for config sources that problem is gone.

Reply:
Having just the bracketed way keeps the actual behavior because `os.ExpandEnv` already attempts to treat everything inside the brackets as an env var (it captures everything until the first `}`). So something like `${HOMEDRIVE:\Windows\Temp}` already doesn't work for env vars and results in an empty string, see https://play.golang.org/p/ucPVlsVDpwZ

Reply:
Keeping the notation without brackets and peeking after the name for ':' and treating it as config source only in that case.